### PR TITLE
[WIP] Refactor chat logic and UI to support multi-chat

### PR DESCRIPTION
**Pull request by pkukielka**

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 1]
```diff
@@ -1,214 +1,144 @@
...
-    val singInWithSourcegraphPanel = SignInWithSourcegraphPanel(project)
     allContentPanel.add(tabbedPane, "tabbedPane", CHAT_PANEL_INDEX)
-    allContentPanel.add(
-        singInWithSourcegraphPanel, SING_IN_WITH_SOURCEGRAPH_PANEL, SIGN_IN_PANEL_INDEX)
-    allContentLayout.show(allContentPanel, SING_IN_WITH_SOURCEGRAPH_PANEL)
-    refreshPanelsVisibility()
+    allContentPanel.add(signInWithSourcegraphPanel, SIGN_IN_PANEL, SIGN_IN_PANEL_INDEX)
+    allContentLayout.show(allContentPanel, SIGN_IN_PANEL)
 
+    ApplicationManager.getApplication().invokeLater { refreshPanelsVisibility() }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: I guess we are in the dispatched thread already so `invokeLater` is redundant (am I right?)

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 2]
```diff
@@ -1,214 +1,144 @@
...
-
-    if (data.isDotcomAccount && data.codyProFeatureFlag) {
-      if (!isSubscriptionTabPresent) {
-        tabbedPane.insertTab(
-            /* title = */ "Subscription",
-            /* icon = */ null,
-            /* component = */ subscriptionPanel,
-            /* tip = */ null,
-            SUBSCRIPTION_TAB_INDEX)
+  private fun runNewCommand(project: Project, commandId: CommandId) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: `project` is a private val of this class already - there is no need to pass it here

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 3]
```diff
@@ -1,214 +1,144 @@
...
-      ApplicationManager.getApplication().invokeLater {
-        messagesPanel.removeAll()
-        chatMessageHistory.clearHistory()
-        inProgressChat.abort()
-        addWelcomeMessage()
-        activateChatTab()
-      }
-      chatId = agent.server.chatNew().get()
-    }
+  private fun startNewChat(project: Project) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: `project` is a private val of this class already - there is no need to pass it here

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 4]
```diff
@@ -1,214 +1,144 @@
...
-        val chatMessages = chat.messages.map { ChatMessage(it.speaker!!, it.text, it.text) }
-        if (item.replyChatId != null) {
-          val restoredId =
-              agent.server
-                  .chatRestore(ChatRestoreParams(model, chatMessages, item.replyChatId!!))
-                  .get()
-          chat.panelId = restoredId
-          chatId = restoredId
+  private fun startNewPanel(
+      project: Project,
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: `project` is a private val of this class already - there is no need to pass it here

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 5]
```diff
@@ -230,176 +160,29 @@ class CodyToolWindowContent(private val project: Project) {
...
     const val SIGN_IN_PANEL_INDEX = 1
     const val ONBOARDING_PANEL_INDEX = 2
-    var logger = Logger.getInstance(CodyToolWindowContent::class.java)
-    const val SING_IN_WITH_SOURCEGRAPH_PANEL = "singInWithSourcegraphPanel"
+
     private const val CHAT_TAB_INDEX = 0
-    private const val RECIPES_TAB_INDEX = 1
-    private const val HISTORY_TAB_INDEX = 2 // todo chat history should be after chat
+    private const val HISTORY_TAB_INDEX = 1
+    private const val RECIPES_TAB_INDEX = 2
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: nit: `s/RECIPES/COMMANDS/`

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/SingleMessagePanel.kt [hunk 6]
```diff
@@ -1,21 +1,23 @@
...
 import org.commonmark.node.Node
 import org.commonmark.parser.Parser
 import org.commonmark.renderer.html.HtmlRenderer
+import java.awt.Color
+import java.util.*
+import javax.swing.JEditorPane
+import javax.swing.JPanel
 
-class MessagePanel(
+class SingleMessagePanel(
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/SingleMessagePanel.kt:20]: why is it "Single"?

File: src/main/java/com/sourcegraph/cody/SubscriptionTabUtils.kt [hunk 7]
```diff
@@ -17,35 +16,30 @@ data class SubscriptionTabPanelData(
 fun fetchSubscriptionPanelData(
     project: Project,
     server: CodyAgentServer
-): CompletableFuture<SubscriptionTabPanelData?> {
+): SubscriptionTabPanelData? {
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/SubscriptionTabUtils.kt:19]: this is correct - my wrapping was redundant 👍

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/SingleMessagePanel.kt (hunk 6 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/chat/ui/SingleMessagePanel.kt:20]: I could just leave it named `MessagePanel` but there is also `MessagesPanel` class (maybe I should name that one other way) and for my taste it was too easy to mistake those two. I wanted to stress the fact that this class wraps just a single message and it's not a panel which contains more of them.

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 8]
```diff
@@ -1,214 +1,128 @@
...
+    CodyAgentService.getInstance(project).onStartup { agent ->
+      agent.client.onNewMessage = Consumer { params ->
+        chatSessions
+            .find { it.getPanelId().getNow(null) == params.id }
+            ?.receiveMessage(params.message)
+      }
 
-          ApplicationManager.getApplication().invokeLater { refreshSubscriptionTab(it) }
+      chatSessions.forEach { session ->
+        val model = "openai/gpt-3.5-turbo"
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: I will encapsulate this into function inside `ChatSession`

File: src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java [hunk 9]
```diff
@@ -72,25 +57,14 @@ public CompletableFuture<Void> webviewCreate(WebviewCreateParams params) {
...
-      } else {
-        logger.warn("webview/postMessage: no messages in transcript");
-      }
+    if (onNewMessage != null
+        && extensionMessage.getType().equals(ExtensionMessage.Type.TRANSCRIPT)) {
+      ApplicationManager.getApplication().invokeLater(() -> onNewMessage.accept(params));
     } else {
       logger.warn("onChatUpdateMessageInProgress is null or message type is not transcript");
-      logger.warn(String.format("webview/postMessage %s: %s", params.getId(), params.getMessage()));
+      logger.warn(String.format("webview/postMessage %s: %s", params.getId(), extensionMessage));
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java]: I `runIde` for a while. The logging here should not cause any trouble (in the scope of #315). However, we will need to reconsider the handling of other types someday.

**Comment by pkukielka** [comment]
Replaced by https://github.com/sourcegraph/jetbrains/pull/365

**Pull request closed without merging**